docs/workflow/tutorial-web-implement-third-party-qr-code-form-element.mdx

```diff
@@ -0,0 +1,114 @@
+
+This article will walk you through creating a form element that uses the [qrcode](https://github.com/soldair/node-qrcode) library to generate and display a QR code.
+
+<img src={useBaseUrl("img/workflow-web-custom-qr-code.png")} />
+
+## Prerequisites
+
+<WebPrereqs />
+
+## Overview
+
+Implementing a QR code custom form element for web applications using a third party library consists of the following steps:
+
+1. Adding third party library dependencies to your project.
+1. Creating the custom form element.
+1. Implementing the custom form element using the third party library.
+
+## Adding Dependencies
+
+1. Run `npm install qrcode` in the terminal to install the third party [qrcode](https://github.com/soldair/node-qrcode) library that generates QR codes.
+1. Run `npm install @types/qrcode` in the terminal to install type information for the third party library. This is optional, but it provides an improved developer experience.
+
+## Set up Custom Form Element Skeleton
+
+To create a new form element:
+
+1. Open the Workflow activity SDK in Visual Studio Code.
+1. Run `npm run generate` in the terminal.
+1. When prompted, select `Form Element`.
+1. Enter the name of the form element you would like to create and press `Enter`. For example, `QrCode`.
+1. Open the newly created `src/activities/QrCode.tsx` file.
+
+## Implement the Custom Form Element
+
+Modify the skeleton form element implementation in `src/activities/QrCode.tsx` to match the following example.
+
+```tsx title="src/activities/QrCode.tsx"
+import React, { useEffect, useState } from "react";
+import QRCode from "qrcode";
+import type { CustomFormElementProps } from "@geocortex/workflow/runtime/app/RegisterCustomFormElementBase";
+import { RegisterCustomFormElementBase } from "@geocortex/workflow/runtime/app/RegisterCustomFormElementBase";
+
+/**
+ * A simple React Component that displays a QR code.
+ * @param props The props that will be provided by the Workflow runtime.
+ */
+const QrCode = (props: CustomFormElementProps) => {
+    const value = props.element.value as string;
+    const [qrCodeUrl, setQrCodeUrl] = useState("");
+
+    // Update the QR code when the value changes
+    useEffect(() => {
+        // Clear the previous QR code
+        setQrCodeUrl("");
+        if (value) {
+            // Asynchronously generate the QR code data URL from the value
+            let didCancel = false;
+            (async () => {
+                const dataUrl = await QRCode.toDataURL(value);
+                if (didCancel) {
+                    return;
+                }
+                setQrCodeUrl(dataUrl);
+            })();
+            return () => {
+                didCancel = true;
+            };
+        }
+    }, [value]);
```

Comment:
😢 this really highlights how much some of the suggestions we talked about (form state mgmt.) could help improve this experience. This pattern seems to come up often.